src/runlib.rs

```diff
@@ -1 +1,375 @@
+            )))
+        }
+    };
+    let stderr = match String::from_utf8(output.stderr) {
+        Ok(output) => output,
+        Err(error) => {
+            return Err(Error::from(io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("Utf8Error: {}", error),
+            )))
+        }
+    };
+    let status = match output.status.code() {
+        Some(code) => code.to_string(),
+        None => "Process terminated by signal".to_string(),
+    };
+
+    byproducts.insert("stdout".to_string(), stdout);
+    byproducts.insert("stderr".to_string(), stderr);
+    byproducts.insert("return-value".to_string(), status);
+
+    Ok(byproducts)
+}
+
+// TODO: implement default trait for in_toto_run's parameters
+
+/// in_toto_run is a function that executes commands on a software supply chain step
+/// (layout inspection coming soon), then generates and returns its corresponding Link metadata.
+/// If a symbolic link cycle is detected in the material or product paths, it is skipped.
+pub fn in_toto_run(
+    name: &str,
+    run_dir: Option<&str>,
+    material_paths: &[&str],
+    product_paths: &[&str],
+    cmd_args: &[&str],
+    key: Option<PrivateKey>,
+    hash_algorithms: Option<&[&str]>,
+    // env: Option<BTreeMap<String, String>>
+) -> Result<Metablock<Json, LinkMetadata>> {
+    // Record Materials: Given the material_paths, recursively traverse and record files in given path(s)
+    let materials = record_artifacts(material_paths, hash_algorithms)?;
+
+    // Execute commands provided in cmd_args
+    let byproducts = run_command(cmd_args, run_dir)?;
+
+    // Record Products: Given the product_paths, recursively traverse and record files in given path(s)
+    let products = record_artifacts(product_paths, hash_algorithms)?;
+
+    // Create link based on values collected above
+    let link_metadata_builder = LinkMetadataBuilder::new()
+        .name(name.to_string())
+        .materials(materials)
+        .byproducts(byproducts)
+        .products(products);
+
+    // TODO Sign the link with key param supplied. If no key param supplied, build & return link
+    // If no key is found, return Metablock with no signatures (for inspection purposes)
+    match key {
+        Some(k) => link_metadata_builder.signed::<Json>(&k),
+        None => link_metadata_builder.unsigned::<Json>(),
+    }
+}
+
+// Helper functions specific to the runlib goes here
+
+/// dir_entry_to_path is a function that, given a DirEntry, return the entry's path as a String
+/// wrapped in Result. If the entry's path is invalid, error is returned.
+fn dir_entry_to_path(
+    entry: std::result::Result<walkdir::DirEntry, walkdir::Error>,
+) -> Result<String> {
+    let path = match entry {
+        Ok(dir_entry) => match dir_entry.path().to_str() {
+            Some(str) => String::from(str),
+            None => {
+                return Err(Error::Programming(format!(
+                    "Invalid Path {}; non-UTF-8 string",
```

Comment:
Is this a valid usage of `Error::Programming` or should I create a specific error for this situation?

Reply:
Gotcha. I looked at the errors and the one that came closest to its meaning ("Walkdir encountered an invalid path format. This might be because of something internal or maybe the paths you supplied were invalid") is `IllegalArgument` error (or maybe `Opaque`), or maybe implementing a new error specific to this use case. Replaced with IllegalArgument, but let me know if something else would be better.